### ISSUE TITLE
Debugger not highlighting current line when using "pipeTransport" in Windows

### ISSUE THREAD

**Issue by Hunter522** [microsoft/vscode-cpptools#470]
I am using "pipeTransport" to remote debug over SSH but the debugger doesn't seem to highlight the current line when I do this in Windows. I've tested this on Linux with the same HelloWorld example and see that it works. See the screenshots as to what I mean.

Some notes:
 - Is this some kind of file path mapping issue?
 - Continue and step into/out/over seem to work and I can see the variables on the left changing correctly, as well as the output looks right in Debug Console 
 - In the Debug Console, I notice that when I hit a breakpoint, the filepath is slightly incorrect with the delimiters: `Breakpoint 1, main () at C:\Users\hnmorgan\Development\odroid-cmake-gcc\src/main.cpp:8`
   - On Linux I see `Breakpoint 1, main () at /home/hmorgan1/Development/odroid-cmake-gcc/src/main.cpp:8`
   - difference here is `\` changes to `/` in the Windows path
 - I am unsure if this is a "pipeTransport" related issue or an issue in general when debugging on Windows
 - Hover debug information doesn't work either

What I see (Windows)
![vscode-cpp-debug-win](https://cloud.githubusercontent.com/assets/9222431/22489554/f389f324-e7cc-11e6-8234-835c76e357c3.png)

What it should look like (Linux)
![vscode-cpp-linux](https://cloud.githubusercontent.com/assets/9222431/22489555/f4c8f2b2-e7cc-11e6-9006-7e16a983d5a7.png)

----

My `launch.json` file on **Windows**: (I've omitted the `attach` section)
```
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "C++ Launch",
            "type": "cppdbg",
            "request": "launch",
            "program": "/root/odroid-cmake-gcc/build/odroid-cmake-gcc",
            "args": [],
            "stopAtEntry": false,
            "cwd": "/root/odroid-cmake-gcc",
            "environment": [],
            "externalConsole": true,
            "pipeTransport": {
                "pipeCwd": "${workspaceRoot}",
                "pipeProgram": "C:\\cygwin64\\bin\\ssh.exe",
                "pipeArgs": ["root@192.168.1.2"],
                "debuggerPath": "/usr/bin/gdb"
            },
            "sourceFileMap": {
                "/root/odroid-cmake-gcc": "C:/Users/hnmorgan/Development/odroid-cmake-gcc"
            },
            
            "linux": {
                "MIMode": "gdb",
                "setupCommands": [
                    {
                        "description": "Enable pretty-printing for gdb",
                        "text": "-enable-pretty-printing",
                        "ignoreFailures": true
                    }
                ]
            },
            "osx": {
                "MIMode": "lldb"
            },
            "windows": {
                "MIMode": "gdb",
                "miDebuggerPath": "C:\\MinGW\\bin\\gdb.exe",
                "setupCommands": [
                    {
                        "description": "Enable pretty-printing for gdb",
                        "text": "-enable-pretty-printing",
                        "ignoreFailures": true
                    }
                ]
            }
        }
    ]
}
```

And here is my `launch.json` on **Linux**:
```
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "C++ Launch",
            "type": "cppdbg",
            "request": "launch",
            "program": "/root/odroid-cmake-gcc/build/odroid-cmake-gcc",
            "args": [],
            "stopAtEntry": false,
            "cwd": "/root/odroid-cmake-gcc",
            "environment": [],
            "externalConsole": true,
            "pipeTransport": {
                "pipeCwd": "${workspaceRoot}",
                "pipeProgram": "/usr/bin/ssh",
                "pipeArgs": ["root@192.168.1.2"],
                "debuggerPath": "/usr/bin/gdb"
            },
            "sourceFileMap": {
                "/root/odroid-cmake-gcc": "${workspaceRoot}"
            },
            
            "linux": {
                "MIMode": "gdb",
                "setupCommands": [
                    {
                        "description": "Enable pretty-printing for gdb",
                        "text": "-enable-pretty-printing",
                        "ignoreFailures": true
                    }
                ]
            },
            "osx": {
                "MIMode": "lldb"
            },
            "windows": {
                "MIMode": "gdb",
                "setupCommands": [
                    {
                        "description": "Enable pretty-printing for gdb",
                        "text": "-enable-pretty-printing",
                        "ignoreFailures": true
                    }
                ]
            }
        }
    ]
}
```

Other information:
 - I am cross-compiling for an ODROID XU3/XU4 platform (arm-linux-gnueabihf), using a Linaro toolchain and CMake to handle building, and using this extension to do the remote debugging over SSH. There are problems with doing remote debugging using gdbserver on the ODROID platform (ODROID-specific issue, not related to this at all) which is why I am doing remote-debugging over SSH.
 - Windows 7 on windows development machine
 - Ubuntu 16.04.1 LTS on linux development machine
 - VSCode 1.8.1
 - C/C++ extension 0.10.0
 - All other extensions disabled

**Comment by pieandcakes** [comment]
@Hunter522 It seems there is a bug in the source mapping. I'll take a look.

**Comment by pieandcakes** [comment]
@Hunter522 Hi, I tried this using cygwin64's ssh to an Ubuntu machine and I can't duplicate this. 

can you add the following to your `launch.json`?
```
"logging":{
    "engineLogging": true,
    "trace": true,
    "traceResponse": true
},
```

We are looking for these two lines:
```
1: (1153) ->*stopped,reason="breakpoint-hit",disp="keep",bkptno="1",frame={addr="0x0000000000400870",func="main",args=[],file="main.cpp",fullname="/src/issues/470/src/main.cpp",line="8"},thread-id="1",stopped-threads="all",core="3"
```
```
E stopped: {"threadId":22560,"reason":"breakpoint","source":{"name":"main.cpp","path":"z:\\Testing\\issues\\470\\src\\main.cpp","sourceReference":0},"line":8,"column":1,"text":null,"allThreadsStopped":true,"type":"stopped"}
```

The top one is what gdb is returning for the stopping event with the fullname of the location.  The second one is what we are sending back to VSCode.

**Comment by Hunter522** [comment]
@pieandcakes  Thanks for looking into this.


Ok added the logging stuff to my `launch.json` and here is what I see:


```
1: (1620) ->*stopped,reason="breakpoint-hit",disp="keep",bkptno="1",frame={addr="0x000106f2",func="main",args=[],file="C:\\Users\\hnmorgan\\Development\\odroid-cmake-gcc\\src/main.cpp",fullname="C:\\Users\\hnmorgan\\Development\\odroid-cmake-gcc\\build/C:\\Users\\hnmorgan\\Development\\odroid-cmake-gcc\\src/main.cpp",line="8"},thread-id="1",stopped-threads="all",core="3"
```

```
E stopped: {"threadId":13772,"reason":"breakpoint","source":{"name":"main.cpp","path":"C:\\Users\\hnmorgan\\Development\\odroid-cmake-gcc\\build\\C:\\Users\\hnmorgan\\Development\\odroid-cmake-gcc\\src\\main.cpp","sourceReference":0},"line":8,"column":1,"text":null,"allThreadsStopped":true,"type":"stopped"}
```


I've also attached my full debug console log here: [debug-console-log.txt](https://github.com/Microsoft/vscode-cpptools/files/748296/debug-console-log.txt)

**Comment by pieandcakes** [comment]
@Hunter522 I looked at the file you sent. I noticed this:

``` 
C setBreakpoints: {"source":{"path":"c:\\Users\\hnmorgan\\Development\\odroid-cmake-gcc\\src\\main.cpp","name":"main.cpp"},"lines":[8,10,12],"breakpoints":[{"line":8},{"line":10},{"line":12}],"sourceModified":false}
1: (943) <-1009-break-insert -f main.cpp:8
1: (948) ->1009^done,bkpt={number="2",type="breakpoint",disp="keep",enabled="y",addr="0x000106f2",func="main()",file="C:\\Users\\hnmorgan\\Development\\odroid-cmake-gcc\\src/main.cpp",fullname="C:\\Users\\hnmorgan\\Development\\odroid-cmake-gcc\\build/C:\\Users\\hnmorgan\\Development\\odroid-cmake-gcc\\src/main.cpp",line="8",thread-groups=["i1"],times="0",original-location="main.cpp:8"}
```

We are sending the right slash but your version of gdb is doing something weird where it has the wrong one in the symbols. Can you check your cmake project and see if it is passing the wrong slashes? 

We get this from gdb:
`1: (1620) ->*stopped,reason="breakpoint-hit",disp="keep",bkptno="1",frame={addr="0x000106f2",func="main",args=[],file="C:\\Users\\hnmorgan\\Development\\odroid-cmake-gcc\\src/main.cpp",fullname="C:\\Users\\hnmorgan\\Development\\odroid-cmake-gcc\\build/C:\\Users\\hnmorgan\\Development\\odroid-cmake-gcc\\src/main.cpp",line="8"},thread-id="1",stopped-threads="all",core="3"
`
 
We are using the full name (which seems to be wrong in this instance). This is the line we send back to the UI. Notice that the path is also wrong. This is coming from the symbols that are generated during compilation.

`
E stopped: {"threadId":13772,"reason":"breakpoint","source":{"name":"main.cpp","path":"C:\\Users\\hnmorgan\\Development\\odroid-cmake-gcc\\build\\C:\\Users\\hnmorgan\\Development\\odroid-cmake-gcc\\src\\main.cpp","sourceReference":0},"line":8,"column":1,"text":null,"allThreadsStopped":true,"type":"stopped"}
`

Based on this, I think its either the way you are compiling or an issue with the version of gdb you are using.

**Comment by Hunter522** [comment]
I see what you mean by GDB reporting incorrect paths from what it sees in the symbols. I tested by building my program in a different directory on host platform, transferring file over to target platform, then debugging it, and I see
``` 
1: (1788) ->*stopped,reason="breakpoint-hit",disp="keep",bkptno="1",frame={addr="0x000106f2",func="main",args=[],file="C:\\Users\\hnmorgan\\Development\\odroid-cmake-gcc\\src/main.cpp",fullname="C:\\Users\\hnmorgan\\Development\\odroid-cmake-gcc-standalone/C:\\Users\\hnmorgan\\Development\\odroid-cmake-gcc\\src/main.cpp",line="8"},thread-id="1",stopped-threads="all",core="0"
```

Note it says `odroid-cmake-gcc-standalone` prefixed now for some reason...(this is the dir I built in, src files live in the other folder)

I will look into this. I'll try a different cross-toolchain if possible. I also compiled with just the cross-toolchain's g++ taking CMake out of the equation, same issue. 

At any rate I think you're right it's either the compiler/linker or GDB on target platform.
